exploration-module/src/main/java/io/edpn/backend/exploration/infrastructure/kafka/KafkaTopicHandler.java

```diff
@@ -0,0 +1,44 @@
+package io.edpn.backend.exploration.infrastructure.kafka;
+
+import lombok.RequiredArgsConstructor;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.ListTopicsResult;
+import org.apache.kafka.clients.admin.NewTopic;
+
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+@RequiredArgsConstructor
+public class KafkaTopicHandler {
+
+    private final AdminClient kafkaAdminClient;
+    private final int topicPartitions;
+    private final short topicReplicationFactor;
+
+
+    public CompletableFuture<String> createTopicIfNotExists(String topicName) {
+        ListTopicsResult listTopicsResult = kafkaAdminClient.listTopics();
+        CompletableFuture<String> topicCreationFuture = new CompletableFuture<>();
+
+        listTopicsResult.names().whenComplete((topicNames, exception) -> {
+            if (exception == null) {
+                if (!topicNames.contains(topicName)) {
+                    NewTopic newTopic = new NewTopic(topicName, topicPartitions, topicReplicationFactor);
+                    kafkaAdminClient.createTopics(Collections.singletonList(newTopic)).all().whenComplete((result, createException) -> {
+                        if (createException == null) {
+                            topicCreationFuture.complete(topicName);
+                        } else {
+                            topicCreationFuture.completeExceptionally(createException);
+                        }
+                    });
+                } else {
+                    topicCreationFuture.complete(topicName);
+                }
+            } else {
+                topicCreationFuture.completeExceptionally(exception);
```

Comment:
I was wondering why this happens in an `else` block, but seems like the Kafka future doesn't quite provide a separate "onError" (or similar) facility. :disappointed:

Considered remarking about the if-elses and nesting here, but it really does seem like a consequence of  lacking an exception hook. I guess one way to make this more pleasing to the eye would be to extract some of it, such as the "create topic if not present" block, to a separate method.